src/mbr.c

```diff
@@ -0,0 +1,270 @@
+				"it's size or twice it's size");
+		goto out;
+	}
+
+	size = boot_part->partition_size;
+
+	/* check if new boot partition overlaps with any other partition */
+	if(mbr_is_part_overlapping(start, size, mbr.partition_table, &ierror)) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	if(!get_sectorsize(fd, &sector_size))
+		sector_size = DEFAULT_SECTOR_SIZE;
+
+	partition->start = (guint64)start * sector_size;
+	partition->size = (guint64)size * sector_size;
+
+	res = TRUE;
+out:
+	if (fd >= 0)
+		g_close(fd, NULL);
+
+	return res;
+}
+
+static void r_mbr_get_chs(guint8 *chs, guint32 sector,
+		const struct hd_geometry *geometry)
+{
+	chs[2] = sector % geometry->sectors + 1;
+	sector /= geometry->sectors;
+	chs[1] = sector % geometry->heads;
+	sector /= geometry->heads;
+	chs[0] = sector & 0xFF;
+	chs[2] |= (sector >> 2) & 0xC0;
+}
+
+static gboolean r_mbr_get_raw_partition_entry(gint fd,
+		struct partition_tbl_entry *raw_entry,
+		const struct mbr_partition *partition, GError **error)
+{
+	gboolean res = FALSE;
+	guint sector_size;
+	struct hd_geometry geometry;
+	GError *ierror = NULL;
+
+	if (!get_sectorsize(fd, &sector_size))
+		sector_size = DEFAULT_SECTOR_SIZE;
+
+	if (partition->start % sector_size || partition->size % sector_size) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Partition start address or size is not a multipe"
+				" of sector size %d", sector_size);
+		goto out;
+	}
+
+	res = get_hd_geometry(fd, &geometry, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	raw_entry->partition_start = partition->start / sector_size;
+	raw_entry->partition_size = partition->size / sector_size;
+
+	r_mbr_get_chs(raw_entry->chs_begin,
+			raw_entry->partition_start,
+			&geometry);
+
+	r_mbr_get_chs(raw_entry->chs_end,
+			raw_entry->partition_start + raw_entry->partition_size - 1,
+			&geometry);
+out:
+	return res;
+}
+
+gboolean r_mbr_set_boot_partition(const gchar *device,
+		const struct mbr_partition *partition,
+		GError **error)
+{
+	gboolean res = FALSE;
+	struct mbr mbr;
+	struct partition_tbl_entry *boot_part;
+	GError *ierror = NULL;
+	gint fd;
+
+	fd = g_open(device, O_RDWR);
+	if (fd == -1) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"opening device failed: %s",
+				g_strerror(errno));
+		goto out;
+	}
+
+	res = r_mbr_read_mbr(fd, &mbr, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(error, ierror,
+				"failed to read MBR:");
+		goto out;
+	}
+
+	boot_part = &mbr.partition_table[BOOT_PARTITION_ENTRY];
+
+	res = r_mbr_get_raw_partition_entry(fd, boot_part, partition, &ierror);
+	if (!res) {
+		g_propagate_prefixed_error(error, ierror,
+				"failed to create new partition entry:");
+		goto out;
+	}
+
+	if (write(fd, &mbr, sizeof(mbr)) != sizeof(mbr)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not write new master boot record: %s",
+				g_strerror(errno));
+		goto out;
+	}
+
```

Comment:
Do we need to tell the kernel to reload the partition table or is that done automatically?

Reply:
This is done automatically.